flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/proxy/DynamoDbStreamsProxy.java

```diff
@@ -0,0 +1,184 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.dynamodb.source.proxy;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.connector.dynamodb.source.split.StartingPosition;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.services.dynamodb.model.DescribeStreamRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeStreamResponse;
+import software.amazon.awssdk.services.dynamodb.model.ExpiredIteratorException;
+import software.amazon.awssdk.services.dynamodb.model.GetRecordsRequest;
+import software.amazon.awssdk.services.dynamodb.model.GetRecordsResponse;
+import software.amazon.awssdk.services.dynamodb.model.GetShardIteratorRequest;
+import software.amazon.awssdk.services.dynamodb.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.dynamodb.model.Shard;
+import software.amazon.awssdk.services.dynamodb.model.StreamStatus;
+import software.amazon.awssdk.services.dynamodb.streams.DynamoDbStreamsClient;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/** Implementation of the {@link StreamProxy} for DynamoDB streams. */
+@Internal
+public class DynamoDbStreamsProxy implements StreamProxy {
+
+    private final DynamoDbStreamsClient dynamoDbStreamsClient;
+    private final SdkHttpClient httpClient;
+    private final Map<String, String> shardIdToIteratorStore;
+
+    private static final Logger LOG = LoggerFactory.getLogger(DynamoDbStreamsProxy.class);
+
+    public DynamoDbStreamsProxy(
+            DynamoDbStreamsClient dynamoDbStreamsClient, SdkHttpClient httpClient) {
+        this.dynamoDbStreamsClient = dynamoDbStreamsClient;
+        this.httpClient = httpClient;
+        this.shardIdToIteratorStore = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public List<Shard> listShards(String streamArn, @Nullable String lastSeenShardId) {
+        return this.getShardsOfStream(streamArn, lastSeenShardId);
+    }
+
+    @Override
+    public GetRecordsResponse getRecords(
+            String streamArn, String shardId, StartingPosition startingPosition) {
+        String shardIterator =
+                shardIdToIteratorStore.computeIfAbsent(
+                        shardId, (s) -> getShardIterator(streamArn, s, startingPosition));
+
+        if (shardIterator == null) {
+            return GetRecordsResponse.builder()
+                    .records(Collections.emptyList())
+                    .nextShardIterator(null)
+                    .build();
+        }
```

Comment:
This is not tested atm.

<img width="490" alt="image" src="https://github.com/apache/flink-connector-aws/assets/35062175/aa3803c4-e25d-46c0-8f3f-a673bdd60ffd">